### PR TITLE
feat: add kurtosis engine logs command that dumps logs for all engines in target dir

### LINKED ISSUE #306: [FR]: Build `kurtosis engine logs` to dump the logs of Kurtosis
### Background & current behavior

During [a user experience session today](https://www.notion.so/kurtosistech/Cam-Sinko-Kurtosis-649a527030764b96bde916b55453b69e?pvs=4), a user's engine was crashing and we weren't sure why. We had to direct the user to dig into Docker commands to figure it out.

We had the same issue in the past as well, about a month ago (I believe @leoporoli was the one debugging it)

### Desired behavior

`kurtosis engine logs` dumps the logs for all the existing engine containers

### How important is this feature or improvement to you?

Painful, the lack of this feature makes using Kurtosis frictionful.

### PR DESCRIPTION
**Pull request by h4ck3rk3y**
## Description:
users don't have to use docker to get engine logs anymore

## Is this change user facing?
YES

## References (if applicable):
Closes https://github.com/kurtosis-tech/kurtosis/issues/306

### CODE REVIEW

**Comment by h4ck3rk3y** [comment]
https://user-images.githubusercontent.com/6670312/227220470-e07cdb23-0e5e-4635-b005-5ea1312b1dfe.mov

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/shared_helpers/shared_helpers.go [hunk 1]
```diff
@@ -459,6 +476,69 @@ func GetEngineAndLogsComponentsNetwork(
 	return targetNetwork, nil
 }
 
+func DumpContainers(ctx context.Context, dockerManager *docker_manager.DockerManager, containers []*types.Container, outputDirpath string) error {
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/shared_helpers/shared_helpers.go:479]: copied this over from container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/docker_kurtosis_backend_enclave_functions.go

File: container-engine-lib/lib/backend_interface/kurtosis_backend.go [hunk 2]
```diff
@@ -63,6 +63,8 @@ type KurtosisBackend interface {
 		resultErr error, // Represents an error with the function itself, rather than the engines
 	)
 
+	EngineLogs(ctx context.Context, outputDirpath string) error
```
Comment (victorcolombo) [inline container-engine-lib/lib/backend_interface/kurtosis_backend.go]: This should start with a verb `GetEngineLogs`

File: cli/cli/commands/engine/engine.go [hunk 3]
```diff
@@ -22,4 +23,5 @@ func init() {
 	EngineCmd.AddCommand(status.StatusCmd)
 	EngineCmd.AddCommand(stop.StopCmd)
 	EngineCmd.AddCommand(restart.RestartCmd)
+	EngineCmd.AddCommand(logs.EngineLogsCmd.MustGetCobraCommand())
```
Comment (victorcolombo) [inline cli/cli/commands/engine/engine.go:26]: Why `MustGetCobraCommand` is here only for logs?

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/engine_functions/engine_logs.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// Permissions for the files & directories we create
+)
+
+func EngineLogs(ctx context.Context, outputDirpath string, dockerManager *docker_manager.DockerManager) error {
+	engineContainerSearchLabels := map[string]string{
+		label_key_consts.AppIDDockerLabelKey.GetString():         label_value_consts.AppIDDockerLabelValue.GetString(),
+		label_key_consts.ContainerTypeDockerLabelKey.GetString(): label_value_consts.EngineContainerTypeDockerLabelValue.GetString(),
+		// NOTE: we do NOT use the engine GUID label here, and instead do postfiltering, because Docker has no way to do disjunctive search!
+	}
+	allEngineContainers, err := dockerManager.GetContainersByLabels(ctx, engineContainerSearchLabels, consts.ShouldFetchAllContainersWhenRetrievingContainers)
```
Comment (victorcolombo) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/engine_functions/engine_logs.go:23]: Are we suppose to have more than 1 engine alive at one time?

**Review by victorcolombo — APPROVED** [review]

File: cli/cli/commands/engine/engine.go (hunk 3 above)
Comment (h4ck3rk3y) [reply cli/cli/commands/engine/engine.go:26]: The others are already native Cobra while logs uses the EngineConsumingKurtosisCommand interface

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/engine_functions/engine_logs.go (hunk 4 above)
Comment (h4ck3rk3y) [reply container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/engine_functions/engine_logs.go:23]: No there are dead/stopped engines as well that this filter catches

**Comment by leeederek** [comment]
Hey @h4ck3rk3y , does this work for both stopped and running engines?

**Comment by h4ck3rk3y** [comment]
@leeederek yes

**Pull request merged**